config/hostcredentials/transform.go

```diff
@@ -0,0 +1,76 @@
+// Package hostcredentials implements a TransformationProvider that fetches
+// credentials from the (oddly named) `host-secrets` service and replaces
+// objects of the form: {$hostcredentials: true} with the credentials.
+//
+// The configuration value `hostCredentialsUrls` must be a list of URLS for
+// host-secrets servers, which will be tried in order repeatedly until one
+// succeeds.
+//
+// Note that this transform will need to run before the "hostsecrets" transform.
+package hostcredentials
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/taskcluster/taskcluster-worker/config"
+)
+
+type provider struct{}
+
+func init() {
+	config.Register("hostcredentials", provider{})
+}
+
+func (provider) Transform(cfg map[string]interface{}) error {
+	urls, ok := cfg["hostCredentialsUrls"].([]interface{})
+	if !ok {
+		return errors.New(
+			"Expected 'hostCredentialsUrls' property to hold a list of URLs for host-secrets services")
+	}
+
+	var creds interface{}
+
+FetchLoop:
+	for {
+		for _, url := range urls {
+			log.Printf("Trying host-secrets server %s...", url)
+
+			resp, err := http.Get(url.(string))
```

Comment:
Do use `go-got` for simple HTTP stuff... disclaimer I wrote go-got :)

```go
g := got.New()
res, err := g.Get(url.(string)).Send()
if err != nil {
  ...
}
if err = json.Unmarshal(res.Body, &creds); err != nil {
   ...
}
```

The default go-got will give you:
- Max 5 retries,
- Exponential back-off: 100, 200, 400, 800 ms with 25% randomization
- Retry any 5xx error,
- ErrResponseTooLarge if response body is larger than 10 MiB,
- Timeout at 30 seconds,

If you want to do it in a loop, that's fine... The 5 retries is just quicker than starting with sleep 60s.
You can also change the backoff policy if you don't like it.

The important part is:
- ErrResponseTooLarge if response body is larger than 10 MiB,
- Timeout at 30 seconds,

This provides additional robustness, in case for some reason the HTTP connection just decides to sit there doing nothing... or if for some reason response tries to run the computer out of memory... That won't happen...
Doing this things right with builtin http in golang is tedious, but the extra robustness is lovely.